jax/lax/lax.py

```diff
@@ -56,6 +57,9 @@
 _min = builtins.max
 _reduce = functools.reduce
 
+Array = Any
```

Comment:
I'm not sure, but a refinement might be `Union[RawArray, core.Tracer]` where where the latter is a bit of a black box, and where `RawArray = Union[onp.ndarray, xla.DeviceArray, pxla.ShardedDeviceArray]` or something. WDYT? Any value in refining that beyond the name?

Also, any value in using [`NewType`](https://docs.python.org/3/library/typing.html#newtype) here rather than an alias?

Reply:
I think this makes sense, but as a next step. I'm concerned that defining the Array type will be tricky to do right, and even though `Array = Any` doesn't add any type safety, it does serve as useful documentation.

Reply:
That makes sense to me!

Reply:
I have never used NewType, but I understand that it is really an abstract (sub)type, e.g., you cannot pass an array to NewType. Aliases seem to be the recommended strategy here. One problem I find with aliases is that they are expanded in error messages, so long aliases can bite you there.

Reply:
I don't think aliases are right here, per the [docs](https://docs.python.org/3/library/typing.html#newtype) I linked (especially the comment at the bottom).

If we did something like

```python
Array = NewType('Array', Any)
```

then we couldn't use an `Any` where an `Array` annotation exists (but we could use an `Array` where an `Any` annotation exists). So it's just better type checking.

Reply:
Also, I suspect the `NewType` approach might lead to better error messages along the lines you described!

Reply:
I would suggest defining different types for input and output arrays:

- Anything *castable* to a JAX array, valid for input into LAX functions (e.g., including NumPy arrays and possibly builtin numbers). We might call these `ArrayLike`, to borrow the a term from NumPy.
- Something guaranteed to to be an actual JAX array object of some form, which would be valid for output from an LAX function. This might just be `Array = Union[core.Tracer, xla.DeviceArray]`.

Reply:
I'm going to submit with this as is for now, since it passes both the JAX tests and all Google-internal tests. If we want to refine our definition of `Array` let's do it in a subsequent PR.

Reply:
I do think that there is value in thinking about a convention for annotating things that can be either a regular Python value or a Tracer. This is important because it marks functions that can be used in user-code, perhaps inside JAX transformations. 

One simple option is to use `Union[some_type, Tracer]`. We can alias this to `ValOrTracer` or something like that.

Does this type need to be parameterized by `some_type`? Perhaps not.